cpp/oneapi/dal/algo/decision_forest/backend/gpu/infer_kernel_impl_dpc.hpp

```diff
@@ -0,0 +1,526 @@
+    de::check_mul_overflow(ctx.row_count_, obs_tree_group_response_count);
+    auto [obs_response_list, zero_obs_response_event] =
+        pr::ndarray<Float, 1>::zeros(queue_,
+                                     ctx.row_count_ * obs_tree_group_response_count,
+                                     alloc::device);
+
+    Float* obs_cls_hist_list_ptr = obs_response_list.get_mutable_data();
+
+    auto local_size = ctx.max_local_size_;
+    const cl::sycl::nd_range<2> nd_range = be::make_multiple_nd_range_2d(
+        { ctx.row_block_count_ * local_size, ctx.tree_in_group_count_ },
+        { local_size, 1 });
+
+    cl::sycl::event last_event = zero_obs_response_event;
+    for (Index proc_tree_count = 0; proc_tree_count < tree_count;
+         proc_tree_count += ctx.tree_in_group_count_) {
+        last_event = queue_.submit([&](cl::sycl::handler& cgh) {
+            cgh.depends_on(deps);
+            cgh.depends_on(last_event);
+            cgh.parallel_for(nd_range, [=](cl::sycl::nd_item<2> item) {
+                const Index local_id = item.get_local_id()[0];
+                const Index local_size = item.get_local_range()[0];
+
+                const Index n_groups = item.get_group_range(0);
+                const Index group_id = item.get_group().get_id(0);
+                const Index n_tree_groups = item.get_group_range(1);
+                const Index tree_group_id = item.get_group().get_id(1);
+                const Index tree_id = proc_tree_count + tree_group_id;
+                const Index leaf_mark = -1;
+
+                const Index elem_count = row_count / n_groups + bool(row_count % n_groups);
+
+                const Index ind_start = group_id * elem_count;
+                const Index ind_end =
+                    cl::sycl::min(static_cast<Index>((group_id + 1) * elem_count), row_count);
+
+                if (tree_id < tree_count) {
+                    const Index* tree_ftr_idx = ftr_idx_list_ptr + tree_id * max_tree_size;
+                    const Index* tree_lch_cls = lch_cls_list_ptr + tree_id * max_tree_size;
+                    const Float* tree_ftr_val = ftr_val_list_ptr + tree_id * max_tree_size;
+                    const double* tree_cls_prb =
+                        cls_prb_list_ptr + tree_id * max_tree_size * class_count;
+
+                    bool tree_root_is_split = leaf_mark != tree_ftr_idx[0];
+
+                    for (Index i = ind_start + local_id; i < ind_end; i += local_size) {
+                        std::uint32_t tree_curr_node = 0;
+                        bool node_is_split = tree_root_is_split;
+                        for (; node_is_split > 0;) {
+                            std::uint32_t idx = node_is_split * tree_ftr_idx[tree_curr_node];
+                            std::uint32_t sn = static_cast<std::uint32_t>(
+                                data_ptr[i * column_count + idx] > tree_ftr_val[tree_curr_node]);
+                            tree_curr_node -=
```

Comment:
Is it always non-positive on the first step?

Reply:
sn is 0 or 1, depending on condition, also i simplified calculation of tree_curr_node